README.md
# A tus client

[![Pub Version](https://img.shields.io/pub/v/tus_client_dart)](https://pub.dev/packages/tus_client_dart)
[![Build Status](https://app.travis-ci.com/tomassasovsky/tus_client.svg?branch=master)](https://travis-ci.org/tomassasovsky/tus_client)

---

A tus client in pure dart. [Resumable uploads using tus protocol](https://tus.io/)
Forked from [tus_client](https://pub.dev/packages/tus_client)

> **tus** is a protocol based on HTTP for _resumable file uploads_. Resumable
> means that an upload can be interrupted at any moment and can be resumed without
> re-uploading the previous data again. An interruption may happen willingly, if
> the user wants to pause, or by accident in case of a network issue or server
> outage.

- [A tus client](#a-tus-client)
  - [Usage](#usage)
    - [Using Persistent URL Store](#using-persistent-url-store)
    - [Adding Extra Headers](#adding-extra-headers)
    - [Adding extra data](#adding-extra-data)
    - [Changing chunk size](#changing-chunk-size)
    - [Pausing upload](#pausing-upload)
  - [Example](#example)
  - [Maintainers](#maintainers)

## Usage

```dart
import 'package:cross_file/cross_file.dart' show XFile;

// File to be uploaded
final file = XFile("/path/to/my/pic.jpg");

// Create a client
final client = TusClient(
    Uri.parse("https://master.tus.io/files/"),
    file,
    store: TusMemoryStore(),
);

// Starts the upload
await client.upload(
    onStart:(TusClient client, Duration? estimate){
        // If estimate is not null, it will provide the estimate time for completion
        // it will only be not null if measuring upload speed
        print('This is the client to be used $client and $estimate time');
    },
    onComplete: () {
        print("Complete!");

        // Prints the uploaded file URL
        print(client.uploadUrl.toString());
    },
    onProgress: (double progress, Duration estimate, TusClient client) {
        print("Progress: $progress, Estimated time: ${estimate.inSeconds}");
    },

    // Set this to true if you want to measure upload speed at the start of the upload
    measureUploadSpeed: true,
);
```

### Using Persistent URL Store

This is only supported on Flutter Android, iOS, desktop and web.
You need to add to your `pubspec.yaml`:

```dart
import 'package:path_provider/path_provider.dart';

//creates temporal directory to store the upload progress
final tempDir = await getTemporaryDirectory();
final tempDirectory = Directory('${tempDir.path}/${gameId}_uploads');
if (!tempDirectory.existsSync()) {
    tempDirectory.createSync(recursive: true);
}

// Create a client
final client = TusClient(
    Uri.parse("https://example.com/tus"),
    file,
    store: TusFileStore(tempDirectory),
);

// Start upload
// Don't forget to delete the tempDirectory
await client.upload();
```

### Adding Extra Headers

```dart
final client = TusClient(
    Uri.parse("https://master.tus.io/files/"),
    file,
    headers: {"Authorization": "..."},
);
```

### Adding extra data

```dart
final client = TusClient(
    Uri.parse("https://master.tus.io/files/"),
    file,
    metadata: {"for-gallery": "..."},
);
```

### Changing chunk size

The file is uploaded in chunks. Default size is 512KB. This should be set considering `speed of upload` vs `device memory constraints`

```dart
final client = TusClient(
    Uri.parse("https://master.tus.io/files/"),
    file,
    maxChunkSize: 10 * 1024 * 1024,  // chunk is 10MB
);
```

### Pausing upload

Pausing upload can be done after current uploading in chunk is completed.

```dart
final client = TusClient(
    Uri.parse("https://master.tus.io/files/"),
    file
);

// Pause after 5 seconds
Future.delayed(Duration(seconds: 5)).then((_) =>client.pause());

// Starts the upload
await client.upload(
    onComplete: () {
        print("Complete!");
    },
    onProgress: (double progress, Duration estimate, TusClient client) {
        print("Progress: $progress, Estimated time: ${estimate.inSeconds}");
    },
);
```

### Set up the retry mechanism

It is posible to set up how many times the upload can fail before throw an error to increase robustness of the upload.
Just indicate how many retries to set up the number of attempts before fail, the retryInterval (in seconds) to indicate the time between every retry
and the retryScale (constant by default) to indicate how this time should increase or not between every retry.

``` dart
final client = TusClient(
    Uri.parse("https://master.tus.io/files/"),
    file
    retries: 5,
    retryInterval: 2,
    retryScale: RetryScale.exponential,
);
```

## Example

For an example of usage in a Flutter app (using file picker) see: [/example](https://github.com/tomassasovsky/tus_client/tree/master/example/lib/main.dart)

## Maintainers

- [Nazareno Cavazzon](https://github.com/NazarenoCavazzon)
- [Jorge Rincon](https://github.com/jorger5)
- [Tomás Sasovsky](https://github.com/tomassasovsky)


CHANGELOG.md
## [2.5.0] - Update http 1.0.0

- Update to http 1.0.0

## [2.3.0] - Added measure real time upload speed

- Added Upload Speed measure optional parameter

## [2.2.3] - Added onStart function and using TusFileStore

- Added better cancel upload method.
- Added TusClientBase abstract class.
- Changed ProtocolExceptions to include code as optional parameter.

## [2.2.2] - Change TusClient upload function

- Added onStart function with TusClient as argument.
- Added cancelUpload function.
- Deleted unused variables.
- Correct typing of functions.
- Changed ProtocolException model to separate code from message.
- Added error handling on requests.

## [2.2.1+3] - Added onStart function and using TusFileStore

- Added onStart function after initiating upload.
- Using TusFileStore for saving video locally (fixes resume-upload function).

## [2.2.1+2] - Fixed metadata and better example

- Fixed generateMetadata() function and improved example.

## [2.2.1+1] - Deleted path dependency

- Deleted path package as dependency.

## [2.2.1] - Change TusClient upload function

- Changed TusClient initialization, headers and metadata are passed now through upload function.

## [2.2.0+1] - Use http client again

- Updated dependencies.
- Now passing reference to the current TusClient in the onProgress function.

## [2.2.0] - Use http client again

- We don't use Dio anymore.

## [2.1.0] - HTTP Package updated

- Now the package uses Dio to manage HTTP Requests.
- Estimated time added.
- Chunk size issue with big files and names fixed.

## [2.0.1] - Added Persistent Store

- Users can now use TusFileStore to create persistent state of uploads.

## [1.0.3] - Updating dependencies

- Updating dependencies.
- Migrating to a native dart package.

## [1.0.2] - Fixed issue with not parsing the http port number

- Fixed issue with not parsing the http port number.
- Fixing formatting.

## [1.0.1] - Fixing custom chunk size

- Fixing handling file as chunks correctly.
- Fixing null safety warnings.
- Updating dependencies.

## [1.0.0] - Migrating to null safety

- Making null safe.
- Increasing minimum Dart SDK.
- Fixing deprecated APIs.

## [0.1.3] - Updating dependencies

- Updating dependencies.
- Removing deadcode.

## [0.1.2] - Many improvements

- Fixing server returns partial url & double header.
- Fixing immediate pause even when uploading with large chunks by timing out the future.
- Removing unused exceptions (deadcode).
- Updating dependencies.

## [0.1.1] - Better file persistence documentation

- Have better documentation on using tus_client_file_store.

## [0.1.0] - Web support

- This is update breaks backwards compatibility.
- Adding cross_file Flutter plugin to manage reading files across platforms.
- Refactoring example to show use with XFile on Android/iOS vs web.

## [0.0.4] - Feature request

- Changing example by adding copying file to be uploaded to application temp directory before uploading.

## [0.0.3] - Bug fix

- Fixing missing Tus-Resumable headers in all requests.

## [0.0.2] - Bug fix

- Fixing failure when offset for server is missing or null.

## [0.0.1] - Initial release

- Support for TUS 1.0.0 protocol.
- Uploading in chunks.
- Basic protocol support.
- **TODO**: Add support for multiple file upload.
- **TODO**: Add support for partial file uploads.


test/src/exceptions_test.dart
import 'package:test/test.dart';
import 'package:tus_client_dart/tus_client_dart.dart';

main() {
  test("exceptions_test.ProtocolException", () {
    final err = ProtocolException(400, "Expected HEADER 'Tus-Resumable'");
    expect(
        "$err",
        "ProtocolException: "
            "Expected HEADER 'Tus-Resumable'");
  });

  test("exceptions_test.ProtocolException.response.shouldRetry", () {
    final err = ProtocolException(400, "Expected HEADER 'Tus-Resumable'");
    expect(
        "$err",
        "ProtocolException: "
            "Expected HEADER 'Tus-Resumable'");
  });

  test("exceptions_test.ProtocolException.response.shouldNotRetry", () {
    final err = ProtocolException(400, "Expected HEADER 'Tus-Resumable'");
    expect(
        "$err",
        "ProtocolException: "
            "Expected HEADER 'Tus-Resumable'");
  });
}


example/lib/main.dart
import 'dart:io';

import 'package:cross_file/cross_file.dart' show XFile;
import 'package:file_picker/file_picker.dart';
import 'package:flutter/material.dart';
import 'package:path_provider/path_provider.dart';
import 'package:tus_client_dart/tus_client_dart.dart';
import 'package:url_launcher/url_launcher.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'TUS Client Upload Demo',
      theme: ThemeData(
        primarySwatch: Colors.teal,
      ),
      home: UploadPage(),
    );
  }
}

class UploadPage extends StatefulWidget {
  @override
  _UploadPageState createState() => _UploadPageState();
}

class _UploadPageState extends State<UploadPage> {
  double _progress = 0;
  Duration _estimate = Duration();
  XFile? _file;
  TusClient? _client;
  Uri? _fileUrl;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('TUS Client Upload Demo'),
      ),
      body: SingleChildScrollView(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: <Widget>[
            SizedBox(height: 12),
            Padding(
              padding: EdgeInsets.symmetric(horizontal: 8, vertical: 2),
              child: Text(
                "This demo uses TUS client to upload a file",
                style: TextStyle(fontSize: 18),
              ),
            ),
            Padding(
              padding: const EdgeInsets.all(6),
              child: Card(
                color: Colors.teal,
                child: InkWell(
                  onTap: () async {
                    _file =
                        await _getXFile(await FilePicker.platform.pickFiles());
                    setState(() {
                      _progress = 0;
                      _fileUrl = null;
                    });
                  },
                  child: Container(
                    padding: EdgeInsets.all(20),
                    child: Column(
                      children: <Widget>[
                        Icon(Icons.cloud_upload, color: Colors.white, size: 60),
                        Text(
                          "Upload a file",
                          style: TextStyle(fontSize: 25, color: Colors.white),
                        ),
                      ],
                    ),
                  ),
                ),
              ),
            ),
            Padding(
              padding: const EdgeInsets.all(8),
              child: Row(
                children: <Widget>[
                  Expanded(
                    child: ElevatedButton(
                      onPressed: _file == null
                          ? null
                          : () async {
                              final tempDir = await getTemporaryDirectory();
                              final tempDirectory = Directory(
                                  '${tempDir.path}/${_file?.name}_uploads');
                              if (!tempDirectory.existsSync()) {
                                tempDirectory.createSync(recursive: true);
                              }

                              // Create a client
                              print("Create a client");
                              _client = TusClient(
                                _file!,
                                store: TusFileStore(tempDirectory),
                                maxChunkSize: 512 * 1024 * 10,
                              );

                              print("Starting upload");
                              await _client!.upload(
                                onStart:
                                    (TusClient client, Duration? estimation) {
                                  print(estimation);
                                },
                                onComplete: () async {
                                  print("Completed!");
                                  tempDirectory.deleteSync(recursive: true);
                                  setState(() => _fileUrl = _client!.uploadUrl);
                                },
                                onProgress: (progress, estimate) {
                                  print("Progress: $progress");
                                  print('Estimate: $estimate');
                                  setState(() {
                                    _progress = progress;
                                    _estimate = estimate;
                                  });
                                },
                                uri: Uri.parse(
                                    "https://tusd.tusdemo.net/files/"),
                                metadata: {
                                  'testMetaData': 'testMetaData',
                                  'testMetaData2': 'testMetaData2',
                                },
                                headers: {
                                  'testHeaders': 'testHeaders',
                                  'testHeaders2': 'testHeaders2',
                                },
                                measureUploadSpeed: true,
                              );
                            },
                      child: Text("Upload"),
                    ),
                  ),
                  SizedBox(width: 8),
                  Expanded(
                    child: ElevatedButton(
                      onPressed: _progress == 0
                          ? null
                          : () async {
                              _client!.pauseUpload();
                            },
                      child: Text("Pause"),
                    ),
                  ),
                ],
              ),
            ),
            Stack(
              children: <Widget>[
                Container(
                  margin: const EdgeInsets.all(8),
                  padding: const EdgeInsets.all(1),
                  color: Colors.grey,
                  width: double.infinity,
                  child: Text(" "),
                ),
                FractionallySizedBox(
                  widthFactor: _progress / 100,
                  child: Container(
                    margin: const EdgeInsets.all(8),
                    padding: const EdgeInsets.all(1),
                    color: Colors.green,
                    child: Text(" "),
                  ),
                ),
                Container(
                  margin: const EdgeInsets.all(8),
                  padding: const EdgeInsets.all(1),
                  width: double.infinity,
                  child: Text(
                      "Progress: ${_progress.toStringAsFixed(1)}%, estimated time: ${_printDuration(_estimate)}"),
                ),
              ],
            ),
            if (_progress > 0)
              ElevatedButton(
                onPressed: () async {
                  final result = await _client!.cancelUpload();

                  if (result) {
                    setState(() {
                      _progress = 0;
                      _estimate = Duration();
                    });
                  }
                },
                child: Text("Cancel"),
              ),
            GestureDetector(
              onTap: _progress != 100
                  ? null
                  : () async {
                      await launchUrl(_fileUrl!);
                    },
              child: Container(
                color: _progress == 100 ? Colors.green : Colors.grey,
                padding: const EdgeInsets.all(8.0),
                margin: const EdgeInsets.all(8.0),
                child:
                    Text(_progress == 100 ? "Link to view:\n $_fileUrl" : "-"),
              ),
            ),
          ],
        ),
      ),
    );
  }

  String _printDuration(Duration duration) {
    String twoDigits(int n) => n.toString().padLeft(2, '0');
    final twoDigitMinutes = twoDigits(duration.inMinutes.remainder(60));
    final twoDigitSeconds = twoDigits(duration.inSeconds.remainder(60));
    return '$twoDigitMinutes:$twoDigitSeconds';
  }

  /// Copy file to temporary directory before uploading
  Future<XFile> _getXFile(FilePickerResult? result) async {
    if (result != null) {
      final chosenFile = result.files.first;
      if (chosenFile.path != null) {
        // Android, iOS, Desktop
        return XFile(chosenFile.path!);
      } else {
        // Web
        return XFile.fromData(
          chosenFile.bytes!,
          name: chosenFile.name,
        );
      }
    }
    return XFile('');
  }
}


lib/src/retry_scale.dart
import 'dart:math';

enum RetryScale {
  /// Same time interval between every retry.
  constant,

  /// If interval is n, on every retry the the interval is increased by n.
  /// For example if [retryInterval] is set to 2 seconds, and the [retries] is set to 4,
  /// the interval for every retry is going to be [2, 4, 6, 8]
  lineal,

  /// If interval is n, on every retry the last interval is going to be duplicated.
  /// For example if [retryInterval] is set to 2 seconds, and the [retries] is set to 4,
  /// the interval for every retry is going to be [2, 4, 8, 16]
  exponential;

  Duration getInterval(int retry, int retryInterval) {
    if (retryInterval == 0) return Duration.zero;
    if (retry == 0) return Duration(seconds: retryInterval);
    switch (this) {
      case RetryScale.constant:
        return Duration(seconds: retryInterval);
      case RetryScale.lineal:
        return Duration(seconds: (retry + 1) * retryInterval);
      case RetryScale.exponential:
        return Duration(seconds: retryInterval * pow(2, retry).toInt());
    }
  }
}


lib/src/tus_client_base.dart
import 'dart:convert';

import 'package:cross_file/cross_file.dart';
import 'package:tus_client_dart/tus_client_dart.dart';

abstract class TusClientBase {
  /// Version of the tus protocol used by the client. The remote server needs to
  /// support this version, too.
  final tusVersion = "1.0.0";

  /// The tus server Uri
  Uri? url;

  Map<String, String>? metadata;

  /// Any additional headers
  Map<String, String>? headers;

  /// Upload speed in Mb/s
  double? uploadSpeed;

  /// List of [Server] that are good for testing speed

  TusClientBase(
    this.file, {
    this.store,
    this.maxChunkSize = 512 * 1024,
    this.retries = 0,
    this.retryScale = RetryScale.constant,
    this.retryInterval = 0,
  });

  /// Create a new upload URL
  Future<void> createUpload();

  /// Checks if upload can be resumed.
  Future<bool> isResumable();

  /// Starts an upload
  Future<void> upload({
    Function(double, Duration)? onProgress,
    Function(TusClient, Duration?)? onStart,
    Function()? onComplete,
    required Uri uri,
    Map<String, String>? metadata = const {},
    Map<String, String>? headers = const {},
  });

  /// Pauses the upload
  Future<bool> pauseUpload();

  /// Cancels the upload
  Future<bool> cancelUpload();

  /// Function to be called after completing upload
  Future<void> onCompleteUpload();

  /// Override this method to customize creating file fingerprint
  String? generateFingerprint() {
    return file.path.replaceAll(RegExp(r"\W+"), '.');
  }

  /// Override this to customize creating 'Upload-Metadata'
  String generateMetadata() {
    final meta = Map<String, String>.from(metadata ?? {});

    if (!meta.containsKey("filename")) {
      // Add the filename to the metadata from the whole directory path.
      //I.e: /home/user/file.txt -> file.txt
      meta["filename"] = file.path.split('/').last;
    }

    return meta.entries
        .map((entry) =>
            entry.key + " " + base64.encode(utf8.encode(entry.value)))
        .join(",");
  }

  /// Storage used to save and retrieve upload URLs by its fingerprint.
  final TusStore? store;

  /// File to upload, must be in[XFile] type
  final XFile file;

  /// The maximum payload size in bytes when uploading the file in chunks (512KB)
  final int maxChunkSize;

  /// The number of times that should retry to resume the upload if a failure occurs after rethrow the error.
  final int retries;

  /// The interval between the first error and the first retry in [seconds].
  final int retryInterval;

  /// The scale type used to increase the interval of time between every retry.
  final RetryScale retryScale;

  /// Whether the client supports resuming
  bool get resumingEnabled => store != null;
}


lib/src/client.dart
import 'dart:async';
import 'dart:developer';
import 'dart:io';
import 'dart:math' show min;
import 'dart:typed_data' show Uint8List, BytesBuilder;
import 'package:tus_client_dart/src/retry_scale.dart';
import 'package:tus_client_dart/src/tus_client_base.dart';

import 'exceptions.dart';
import 'package:http/http.dart' as http;

/// This class is used for creating or resuming uploads.
class TusClient extends TusClientBase {
  TusClient(
    super.file, {
    super.store,
    super.maxChunkSize = 512 * 1024,
    super.retries = 0,
    super.retryScale = RetryScale.constant,
    super.retryInterval = 0,
  }) {
    _fingerprint = generateFingerprint() ?? "";
  }

  /// Override this method to use a custom Client
  http.Client getHttpClient() => http.Client();

  int _actualRetry = 0;

  /// Create a new [upload] throwing [ProtocolException] on server error
  Future<void> createUpload() async {
    try {
      _fileSize = await file.length();

      final client = getHttpClient();
      final createHeaders = Map<String, String>.from(headers ?? {})
        ..addAll({
          "Tus-Resumable": tusVersion,
          "Upload-Metadata": _uploadMetadata ?? "",
          "Upload-Length": "$_fileSize",
        });

      final _url = url;

      if (_url == null) {
        throw ProtocolException('Error in request, URL is incorrect');
      }

      final response = await client.post(_url, headers: createHeaders);

      if (!(response.statusCode >= 200 && response.statusCode < 300) &&
          response.statusCode != 404) {
        throw ProtocolException(
            "Unexpected Error while creating upload", response.statusCode);
      }

      String urlStr = response.headers["location"] ?? "";
      if (urlStr.isEmpty) {
        throw ProtocolException(
            "missing upload Uri in response for creating upload");
      }

      _uploadUrl = _parseUrl(urlStr);
      store?.set(_fingerprint, _uploadUrl as Uri);
    } on FileSystemException {
      throw Exception('Cannot find file to upload');
    }
  }

  Future<bool> isResumable() async {
    try {
      _fileSize = await file.length();
      _pauseUpload = false;

      if (!resumingEnabled) {
        return false;
      }

      _uploadUrl = await store?.get(_fingerprint);

      if (_uploadUrl == null) {
        return false;
      }
      return true;
    } on FileSystemException {
      throw Exception('Cannot find file to upload');
    } catch (e) {
      return false;
    }
  }

  /// Start or resume an upload in chunks of [maxChunkSize] throwing
  /// [ProtocolException] on server error
  Future<void> upload({
    Function(double, Duration)? onProgress,
    Function(TusClient, Duration?)? onStart,
    Function()? onComplete,
    required Uri uri,
    Map<String, String>? metadata = const {},
    Map<String, String>? headers = const {},
    bool measureUploadSpeed = false,
  }) async {
    setUploadData(uri, headers, metadata);

    final _isResumable = await isResumable();

    if (!_isResumable) {
      await createUpload();
    }

    // get offset from server
    _offset = await _getOffset();

    // Save the file size as an int in a variable to avoid having to call
    int totalBytes = _fileSize as int;

    // We start a stopwatch to calculate the upload speed
    final uploadStopwatch = Stopwatch()..start();

    // start upload
    final client = getHttpClient();

    if (onStart != null) {
      Duration? estimate;
      if (uploadSpeed != null) {
        final _workedUploadSpeed = uploadSpeed! * 1000000;

        estimate = Duration(
          seconds: (totalBytes / _workedUploadSpeed).round(),
        );
      }
      // The time remaining to finish the upload
      onStart(this, estimate);
    }

    while (!_pauseUpload && _offset < totalBytes) {
      if (!File(file.path).existsSync()) {
        throw Exception("Cannot find file ${file.path.split('/').last}");
      }
      final uploadHeaders = Map<String, String>.from(headers ?? {})
        ..addAll({
          "Tus-Resumable": tusVersion,
          "Upload-Offset": "$_offset",
          "Content-Type": "application/offset+octet-stream"
        });

      await _performUpload(
        onComplete: onComplete,
        onProgress: onProgress,
        uploadHeaders: uploadHeaders,
        client: client,
        uploadStopwatch: uploadStopwatch,
        totalBytes: totalBytes,
      );
    }
  }

  Future<void> _performUpload({
    Function(double, Duration)? onProgress,
    Function()? onComplete,
    required Map<String, String> uploadHeaders,
    required http.Client client,
    required Stopwatch uploadStopwatch,
    required int totalBytes,
  }) async {
    try {
      final request = http.Request("PATCH", _uploadUrl as Uri)
        ..headers.addAll(uploadHeaders)
        ..bodyBytes = await _getData();
      _response = await client.send(request);

      if (_response != null) {
        _response?.stream.listen(
          (newBytes) {
            if (_actualRetry != 0) _actualRetry = 0;
          },
          onDone: () {
            if (onProgress != null && !_pauseUpload) {
              // Total byte sent
              final totalSent = _offset + maxChunkSize;
              double _workedUploadSpeed = 1.0;

              // If upload speed != null, it means it has been measured
              if (uploadSpeed != null) {
                // Multiplied by 10^6 to convert from Mb/s to b/s
                _workedUploadSpeed = uploadSpeed! * 1000000;
              } else {
                _workedUploadSpeed =
                    totalSent / uploadStopwatch.elapsedMilliseconds;
              }

              // The data that hasn't been sent yet
              final remainData = totalBytes - totalSent;

              // The time remaining to finish the upload
              final estimate = Duration(
                seconds: (remainData / _workedUploadSpeed).round(),
              );

              final progress = totalSent / totalBytes * 100;
              onProgress((progress).clamp(0, 100), estimate);
              _actualRetry = 0;
            }
          },
        );

        // check if correctly uploaded
        if (!(_response!.statusCode >= 200 && _response!.statusCode < 300)) {
          throw ProtocolException(
            "Error while uploading file",
            _response!.statusCode,
          );
        }

        int? serverOffset = _parseOffset(_response!.headers["upload-offset"]);
        if (serverOffset == null) {
          throw ProtocolException(
              "Response to PATCH request contains no or invalid Upload-Offset header");
        }
        if (_offset != serverOffset) {
          throw ProtocolException(
              "Response contains different Upload-Offset value ($serverOffset) than expected ($_offset)");
        }

        if (_offset == totalBytes && !_pauseUpload) {
          this.onCompleteUpload();
          if (onComplete != null) {
            onComplete();
          }
        }
      } else {
        throw ProtocolException("Error getting Response from server");
      }
    } catch (e) {
      if (_actualRetry >= retries) rethrow;
      final waitInterval = retryScale.getInterval(
        _actualRetry,
        retryInterval,
      );
      _actualRetry += 1;
      log('Failed to upload,try: $_actualRetry, interval: $waitInterval');
      await Future.delayed(waitInterval);
      return await _performUpload(
        onComplete: onComplete,
        onProgress: onProgress,
        uploadHeaders: uploadHeaders,
        client: client,
        uploadStopwatch: uploadStopwatch,
        totalBytes: totalBytes,
      );
    }
  }

  /// Pause the current upload
  Future<bool> pauseUpload() async {
    try {
      _pauseUpload = true;
      await _response?.stream.timeout(Duration.zero);
      return true;
    } catch (e) {
      throw Exception("Error pausing upload");
    }
  }

  Future<bool> cancelUpload() async {
    try {
      await pauseUpload();
      await store?.remove(_fingerprint);
      return true;
    } catch (_) {
      throw Exception("Error cancelling upload");
    }
  }

  /// Actions to be performed after a successful upload
  Future<void> onCompleteUpload() async {
    await store?.remove(_fingerprint);
  }

  void setUploadData(
    Uri url,
    Map<String, String>? headers,
    Map<String, String>? metadata,
  ) {
    this.url = url;
    this.headers = headers;
    this.metadata = metadata;
    _uploadMetadata = generateMetadata();
  }

  /// Get offset from server throwing [ProtocolException] on error
  Future<int> _getOffset() async {
    final client = getHttpClient();

    final offsetHeaders = Map<String, String>.from(headers ?? {})
      ..addAll({
        "Tus-Resumable": tusVersion,
      });
    final response =
        await client.head(_uploadUrl as Uri, headers: offsetHeaders);

    if (!(response.statusCode >= 200 && response.statusCode < 300)) {
      throw ProtocolException(
        "Unexpected error while resuming upload",
        response.statusCode,
      );
    }

    int? serverOffset = _parseOffset(response.headers["upload-offset"]);
    if (serverOffset == null) {
      throw ProtocolException(
          "missing upload offset in response for resuming upload");
    }
    return serverOffset;
  }

  /// Get data from file to upload

  Future<Uint8List> _getData() async {
    int start = _offset;
    int end = _offset + maxChunkSize;
    end = end > (_fileSize ?? 0) ? _fileSize ?? 0 : end;

    final result = BytesBuilder();
    await for (final chunk in file.openRead(start, end)) {
      result.add(chunk);
    }

    final bytesRead = min(maxChunkSize, result.length);
    _offset = _offset + bytesRead;

    return result.takeBytes();
  }

  int? _parseOffset(String? offset) {
    if (offset == null || offset.isEmpty) {
      return null;
    }
    if (offset.contains(",")) {
      offset = offset.substring(0, offset.indexOf(","));
    }
    return int.tryParse(offset);
  }

  Uri _parseUrl(String urlStr) {
    if (urlStr.contains(",")) {
      urlStr = urlStr.substring(0, urlStr.indexOf(","));
    }
    Uri uploadUrl = Uri.parse(urlStr);
    if (uploadUrl.host.isEmpty) {
      uploadUrl = uploadUrl.replace(host: url?.host, port: url?.port);
    }
    if (uploadUrl.scheme.isEmpty) {
      uploadUrl = uploadUrl.replace(scheme: url?.scheme);
    }
    return uploadUrl;
  }

  http.StreamedResponse? _response;

  int? _fileSize;

  String _fingerprint = "";

  String? _uploadMetadata;

  Uri? _uploadUrl;

  int _offset = 0;

  bool _pauseUpload = false;

  /// The URI on the server for the file
  Uri? get uploadUrl => _uploadUrl;

  /// The fingerprint of the file being uploaded
  String get fingerprint => _fingerprint;

  /// The 'Upload-Metadata' header sent to server
  String get uploadMetadata => _uploadMetadata ?? "";
}
